### PR TITLE
Make ChunkedCELoss support torch.autograd.grad

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #3250
* #3248
* #3247
* #3246
* __->__ #3249

Add a ``support_autograd_grad`` opt-in flag to ``ChunkedCELoss.Config``
that exposes the lm_head parameter gradients as explicit autograd
outputs of the returned loss tensor. Designed for FX-tracing flows
(e.g. graph_trainer's aot_fx_trace) where ``param.grad`` side-effect
writes from ``chunk_loss.backward()`` inside the chunk loop don't
survive into the captured graph and replay therefore produces
all-zero param grads.

Mechanism, when the flag is True:

  - The chunk loop runs unchanged: per-chunk ``chunk_loss.backward()``
    populates ``lm_head.weight.grad`` with the correctly sharded value
    (FSDP last-chunk reduce-scatter still handles the actual reduction).
  - After the loop, the sharded ``param.grad`` is captured via
    ``p.grad.detach()`` and ``p.grad`` is cleared.
  - The captured grads (plus the existing accumulated hidden_states
    grad) are plumbed through a new ``_ChunkedLossWithParamGrads``
    autograd Function as saved tensors. Its backward returns them as
    grads for the corresponding lm_head parameter inputs, so outer
    ``torch.autograd.grad(loss, lm_head.parameters())`` resolves to
    real gradients instead of None / zeros.
  - Under FSDP, ``set_requires_gradient_sync(False)`` is set on lm_head
    and restored after the outer backward via a callback queued on the
    autograd engine. Without this, outer ``loss.backward()`` would
    re-fire the post-accumulate-grad hook on already-sharded grads and
    either error or produce wrong values.

Both autograd Functions (the existing ``_DecoderOutputGradientBackProp``
and the new ``_ChunkedLossWithParamGrads``) return saved grads as-is
without chain-ruling through ``grad_output``. The contract is that the
loss returned by these Functions is the autograd endpoint — callers
must pass any scaling factor (e.g. ``global_valid_tokens``) into
``loss_fn`` rather than dividing the returned loss externally. See
graph_trainer's ``compute_loss`` for the canonical pattern. This
matches the pre-existing behavior of ``_DecoderOutputGradientBackProp``
and avoids a structural FSDP+TP problem: ``grad_output`` is a
``Replicate()`` scalar DTensor on the loss's mesh (typically ``(tp,)``)
while saved param grads live on the params' mesh (e.g. ``(fsdp, tp)``);
DTensor refuses cross-mesh ``aten.mul.Tensor``, so any chain-rule
multiply would crash at runtime.

Tests:

  - tests/unit_tests/test_loss.py: parametrized ``support_autograd_grad
    in {False, True}`` equivalence check against the unchunked CE
    standard path; bitwise (rtol=0, atol=0) check that True and False
    paths produce identical grads; side-effect contract check that the
    True path doesn't populate ``param.grad``.
  - graph_trainer/tests/test_trace_module.py: end-to-end test that
    traces a small ``lm_head + ChunkedCELoss(support_autograd_grad=
    True)`` train_step via ``trace_train_step`` and verifies
    ``torch.equal`` between eager and replayed loss + h grad + lm_head
    grad on a CUDA model.

The flag defaults to False so existing callers (the eager torchtitan
trainer) are unaffected; consumers that want explicit param grads
opt in. graph_trainer wires this in a separate commit.

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -229,6 +229,16 @@ class ChunkedCELoss(BaseLoss):
...
         num_chunks: int = 8
         """Number of chunks to split the sequence into."""
+        support_autograd_grad: bool = False
+        """If True, plumb the sharded lm_head param grads (already produced
+        by the chunk loop's reduce-scatter) out as explicit autograd outputs
+        so outer ``torch.autograd.grad(loss, [hidden_states,
+        *lm_head.parameters()])`` returns them directly — no reliance on
+        ``param.grad`` side effects. Required for FX tracing / replay where
+        side-effecting ``.grad`` writes don't survive. Compatible with both
+        outer ``loss.backward()`` and ``torch.autograd.grad`` consumers.
```
Comment (SherlockNoMad) [inline torchtitan/components/loss.py]: this is a smart fix. 

do you have a tlparse on the graph of chunked loss I can take a look?

Comment (ydwu4) [reply torchtitan/components/loss.py]: Sure here is one: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/yidi-graph-trainer-moe-deepseek-v3-debugmodel-no-chainrule/rank_0/-_-_-_-/make_fx_graph_traced_1.txt?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000


For a smaller model: check https://www.internalfb.com/phabricator/paste/view/P2311018461

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -229,6 +229,16 @@ class ChunkedCELoss(BaseLoss):
     class Config(BaseLoss.Config):
         num_chunks: int = 8
         """Number of chunks to split the sequence into."""
+        support_autograd_grad: bool = False
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: sounds that a cleaner, non-intrusive way would be
- making a subclass of chunked loss class, e.g. by turning the autograd function a class variable and override it
- swapping to use it for graph trainer configs

Comment (ydwu4) [reply torchtitan/components/loss.py]: Sure, refactored accordingly!

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -345,15 +345,64 @@ def __call__(
...
+        returns ``accumulated_grad`` as the gradient for ``hidden_states``,
+        propagating through the decoder. Subclasses override to swap in a
+        different autograd Function.
+        """
         return _DecoderOutputGradientBackProp.apply(
             hidden_states, accumulated_grad, total_loss
         )
 
 
+class ChunkedCELossWithParamGrads(ChunkedCELoss):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: please define this in graph trainer, as it's not required as part of core

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -389,4 +438,86 @@ def backward(  # pyrefly: ignore[bad-override]
...
         # but expressed as a return value so autograd handles the traversal
         # in a single pass (no "backward through graph twice" error).
+        # Note: this is not safe if downstream accidentally runs tensor ops after
+        # the loss returns, which would produce a non-trivial grad_output that we need
+        # to properly handle. The complicated part is that grad_output might not be
+        # on the same device mesh as accumlated_grad.
         return accumulated_grad, None, None
+
+
+class _ChunkedLossWithParamGrads(torch.autograd.Function):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: same here

File: torchtitan/components/loss.py (hunk 3 above)
Comment (ydwu4) [reply torchtitan/components/loss.py]: Moved into graph trainer.

**Review by tianyu-l — APPROVED** [review]
lgtm on the core side

File: torchtitan/components/loss.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/components/loss.py]: nice, the graph looks good to me.

File: torchtitan/experiments/graph_trainer/chunked_loss.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    Designed for graph_trainer, where the chunk loop's per-chunk
+    ``param.grad`` side-effect writes don't survive the captured graph and
+    replay therefore produces all-zero param grads. Compatible with both
+    outer ``loss.backward()`` and ``torch.autograd.grad`` consumers.
+    """
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(ChunkedCELoss.Config):
+        pass
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/chunked_loss.py]: not sure if this required or not?

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
in the latter PR, we should have a variant that uses chunked loss in the test_bitwise_deterministic.py

File: torchtitan/experiments/graph_trainer/chunked_loss.py (hunk 5 above)
Comment (ydwu4) [reply torchtitan/experiments/graph_trainer/chunked_loss.py]: Probably not. Can just remove it I feel. If we need, we could add it back later.
